vignettes/test-children.Rmd

```diff
@@ -0,0 +1,165 @@
+
+When testing code that relies on an API, there are two broad categories of testing:
+
+- **Integration testing** - The tests call to the server each time they are run. 
+  While this is nice in theory, it can cause problems because test code can
+  overload the servers because of how frequently they need to run which will not only
+  cause your tests to fail, but negatively affect all users of the API.
+- **Mock testing** - Rather than actually call the server, the tests use "mocks"
+  of what should be returned by the server. This allows tests to be run more 
+  quickly and without stressing the server. The hard part is creating these mocks,
+  but rcrunch has tools that try to make this as easy as possible.
+
+
+## Integration testing strategy
+We do not advise that child packages run integration tests. If you feel you need them, 
+please contact the Crunch team and we may be able to grant you access to the 
+Crunch.io test server and help you write tests that use that.
+
+
+## Mock testing strategy
+Mock tests for the crunch package are tested using the `httptest` package, which was 
+written while the author worked at Crunch.io and so is tailored to our situation quite
+well. Before continuing, I highly recommend reading the vignettes for that package:
+
+- [`httptest`: A Test Environment for HTTP Requests in R](https://enpiar.com/r/httptest/articles/httptest.html)
+- [Redacting and Modifying Recorded Requests](https://enpiar.com/r/httptest/articles/redacting.html)
+
+
+### Step 1: Capture requests (ideally, reproducibly)
+The `httptest` package has functions `start_capturing()` and `stop_capturing()`
+that allow it to save the results of API calls that you make in 
+a way that it will be able to replay later, while you're testing. You'll
+want to capture all the requests needed to run the test and save them
+in the `testthat` folder of your package so that your tests have access to
+them. If you've already written the test & function, you can run that same
+code inside of `httptest`'s capturing mechanism. However, if you're trying
+to write your tests before you've written the function, you'll need to 
+plan out what `crunch` functions you will need to run so that you can
+capture them.
+
+It is a really good idea to save this code in a reproducible state
+so that you can rerun it if your code changes, or if the crunch API 
+changes. To make it reproducible, I recommend:
+
+1) Saving the script in your package, a good place is in the `testthat`
+   folder, at the level above where the fixtures are saved.
+2) Deleting existing mock files if they already exist.
+3) Either creating the dataset in the script, or at least starting from 
+   a dataset in a project stored in a place that everyone who works on
+   the package have access to.
+
+Putting that altogether, here's an example of a script. It could be saved
+for example in a file `"tests/testthat/captured/generate_1.R"`.
+```{r, eval = FALSE}
+library(crunch)
+library(httptest)
+
```

Comment:
Let's add an explicit `rm(list=ls())` here. I've found weird stuff happens if I don't clear my workspace appropriately. We should also make note that a restart of R + clearing workspace is recommended before capturing a test.